### ISSUE TITLE
Suggestions on how we can improve django-rq and add support to Django 4

### ISSUE THREAD

**Issue by rlp1f22** [rq/django-rq#528]
Hello guys.

First of all, I wanna say that this lib is awesome. So, thank you!

The reason I'm sending this question is to get some feedback from you guys, and I'm also hoping to provide some insight on how we can make django-rq better. I don't know if I'll be able to do the latter since I'm sort of a noob, but I'll give it a shot.

Django 4 has added support for caching with Redis:
- [redis-cache-backend](https://docs.djangoproject.com/en/4.0/releases/4.0/#redis-cache-backend)
- [Cache ~ Redis](https://docs.djangoproject.com/en/4.0/topics/cache/#redis)

I know not everyone can take advantage of this right away, but this might be helpful for people like me who are starting a new project today.

So, after reading the source code [django/core/cache/backends/redis.py](https://github.com/django/django/blob/stable/4.0.x/django/core/cache/backends/redis.py), I can say that the integration django<->redis-py is quite elegant. I'm not 100% sure, and that's why I'm hoping to get some feedback from you guys, but this new cache backend kind of solves some issues.

For instance, [issue #210](https://github.com/rq/django-rq/issues/210), _get_redis_connection() creates new redis client every time it is called_, and according to [redis-py ~ Connection Pools](https://github.com/redis/redis-py#connection-pools): 

> By default, each Redis instance you create will in turn create its own connection pool.

I know we can just:
```python
redis_cursor = redis.StrictRedis(host='', port='', db='', password='')
high_queue = django_rq.get_queue('high', connection=redis_cursor)
```
In order to avoid creating new connections and controlling client instances.

> Small tangent:
> 
> Client Classes: Redis and StrictRedis
> redis-py 3.0 drops support for the legacy Redis client class. StrictRedis has been renamed to Redis and an alias named StrictRedis is provided so that users previously using StrictRedis can continue to run unchanged.

However, we can make this easier by using Django's new backend:
- [Django ~ Low level cache API](https://docs.djangoproject.com/en/4.0/topics/cache/#the-low-level-cache-api)
- [RedisCacheClient & RedisCache](https://github.com/django/django/blob/stable/4.0.x/django/core/cache/backends/redis.py)

```python
from django.core.cache import caches

my_cache = caches["name_of_my_redis_backend"]
redis_client = my_cache._cache
redis_client.get_client()
# ...
```

Note that the `_cache` method has the  [cached_property](https://docs.djangoproject.com/en/4.0/ref/utils/#django.utils.functional.cached_property) decorator, which, as far as I can tell, will return the same instance every time it's called. This seems nicer than dealing with a singleton and instantiating a redis client directly like `r = redis.Redis(...)`. Django already takes care of almost everything via the Cache API.

```python
class RedisCache(BaseCache):
    def __init__(self, server, params):
        # ...
        self._class = RedisCacheClient
        # ...

    @cached_property
    def _cache(self):
        return self._class(self._servers, **self._options)
```

Here's how my settings file looks:

```python
# settings.py

CACHES = { 
    "default": {
        "BACKEND": "django.core.cache.backends.redis.RedisCache",
        "LOCATION": REDIS_LOCATION,
        "OPTIONS": {
            "db": "0",
            "parser_class": "redis.connection.HiredisParser",
            "pool_class": "redis.BlockingConnectionPool",
        },
        "TIMEOUT": 300,
    }   
}
```

My point is: getting a client connection from Django's cache API is probably simpler. Most of the details (connection pools, redis client instances, etc.) have already been taken care of. Even when considering backwards compatibility, we can still change some stuff to mimic some parts of django's redis backend.

Also, the `get_redis_connection()` function from the queues module does not work with Django's new redis backend *IF* we try to use the  `USE_REDIS_CACHE` config.

```python
    if 'USE_REDIS_CACHE' in config.keys():
        # ...

        from django.core.cache import caches
        cache = caches[config['USE_REDIS_CACHE']]
        # We're using django-redis-cache
        try:
            return cache._client
        except AttributeError:
            # For django-redis-cache > 0.13.1
            return cache.get_master_client()
```

Fixing this to add support for Django 4 would be pretty simple since we can just do `caches["name_of_my_redis_backend"]._cache.get_client(...)`. Everything else could remain the same to maintain compatibility. 

Another thing that caught my attention was this: [persistent database connections? #228](https://github.com/rq/django-rq/issues/228).

It seems Django-rq uses the `reset_db_connections()` function to close all database connections via `django.db.connections` --  [rqworker.py](https://github.com/rq/django-rq/blob/master/django_rq/management/commands/rqworker.py).

Why is this necessary? I mean, let's say I'm running some Docker containers, one for each -- Redis, Postgres, and Django --, why must I close all database connections? Why do postgres connections impact redis connections? 

I don't fully understand everything that Django-rq (or even RQ) does, so I'm guessing that there's a good reason for that. Could you guys clue me in?

@selwin mentioned:
> Not at the moment, Django-RQ explicitly closes all DB connections prior to calling os.fork() and executing jobs.
> Suggestions on how we can make this more efficient is welcome.

But if you guys could elaborate a bit.. it would be nice.

Django's redis backend also handles connection pools via the `_get_connection_pool()` method when we `get_client()`. So, this seems like it could improve the overall architecture of Django-rq? Again, I'm not 100% sure... so I might be saying something really dumb, and that's why I wanna get some feedback.

---
---

I'm currently working on a "stripped-down" version of django-rq. I don't need sentry support, statistics, admin interface, and I think I can simplify some stuff by using Django's new cache backend. This only applies to my current requirements, though. I might need all of that in the future. 

I'd love to share what I have so far, but I kinda wanna get your take on some of the stuff I mentioned above. I was also hoping to contribute to this project, mainly by adding support to Django 4 and addressing some of the stuff above.

If you guys think this is a good idea, I can open a pull request this week or the next. 

Thank you!

**Comment by DragosDumitrache** [comment]
I'm just butting in, as I've been using this with Django 4, but what kind of error would we be seeing? Asking, as I keep encountering an improperly configured database error, saying I need to pass in `engine`, despite my database being configured properly

**Comment by selwin** [comment]
> My point is: getting a client connection from Django's cache API is probably simpler. Most of the details (connection pools, redis client instances, etc.) have already been taken care of. Even when considering backwards compatibility, we can still change some stuff to mimic some parts of django's redis backend.

Yes, I'm all for this. If someone can open a PR to support using the same Redis connection from Django's cache backend, I'm happy to review and merge it in.
